### PR TITLE
Switch to Cabal 2.2 (prerelease)

### PR DESCRIPTION
**Pull request by snoyberg**
Note: Documentation fixes for https://docs.haskellstack.org/en/stable/ should target the "stable" branch, not master.

Please include the following checklist in your PR:

* [X] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [X] The documentation has been updated, if necessary.

CC @phadej 

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: src/Stack/Package.hs [hunk 1]
```diff
@@ -129,14 +127,14 @@ rawParseGPD
...
+      Left (mversion, errs) -> throwM $ PackageInvalidCabalFile key
+        (fromCabalVersion <$> mversion)
+        errs
+        warnings
+      Right gpkg -> return (warnings, gpkg)
   where
-    chars = T.unpack (dropBOM (decodeUtf8With lenientDecode bs))
-
-    -- https://github.com/haskell/hackage-server/issues/351
-    dropBOM t = fromMaybe t $ T.stripPrefix "\xFEFF" t
```
Comment (snoyberg) [inline src/Stack/Package.hs]: @phadej Can we rely on Cabal stripping out the byte-order marker from the bytestring, or do we need to do something more intelligent here?

Comment (phadej) [reply src/Stack/Package.hs]: Cabal strips BOM (and warns about, if sees it). `parseGenericPackageDescription` also verifies that `ByteString` looks like valid UTF8, so you don't to do that either.

TL;DR you don't need to do any preprocessing.

Comment (snoyberg) [reply src/Stack/Package.hs]: Awesome, thanks!

**Comment by mgsloan** [comment]
LGTM :+1:

**Comment by snoyberg** [comment]
After fighting with version bounds in a few different places, this is not blocked on:

https://github.com/fizruk/http-api-data/issues/72

The issue: http-api-data has a `Cabal < 2.2` unnecessarily in its `custom-setup` stanza. The maintainer wants to wait until `blaze-builder` adds GHC 8.4 support before relaxing that bound. So our choices are:

* Include `allow-newer` in the `stack.yaml` files, making the released version basically unbuildable directly from Hackage
* Hold off indefinitely on the upgrade
* Try to push ahead with dropping http-api-data from the dependency chain of persistent (https://github.com/yesodweb/persistent/issues/800)
* Have Cabal as a submodule and/or a forked version on Hackage, and only pull in the relevant parsing code we need, bypassing the erroneous upper bounds

I kind of like the last option, as it makes us more disconnected from the Cabal release process over all, but it adds some maintenance overhead. And from my initial experience, there's unfortunately a very large module graph involved in the parsing code.
